ksql-rest-app/src/main/java/io/confluent/ksql/rest/util/CommandTopic.java

```diff
@@ -0,0 +1,173 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.ksql.rest.util;
+
+import com.google.common.collect.Lists;
+import io.confluent.kafka.serializers.KafkaJsonDeserializer;
+import io.confluent.kafka.serializers.KafkaJsonDeserializerConfig;
+import io.confluent.kafka.serializers.KafkaJsonSerializer;
+import io.confluent.ksql.rest.server.computation.Command;
+import io.confluent.ksql.rest.server.computation.CommandId;
+import io.confluent.ksql.rest.server.computation.QueuedCommand;
+import io.confluent.ksql.rest.server.computation.QueuedCommandStatus;
+import io.confluent.ksql.rest.server.computation.RestoreCommands;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CommandTopic {
+
+  private static final Logger log = LoggerFactory.getLogger(CommandTopic.class);
+
+  private final Consumer<CommandId, Command> commandConsumer;
+  private final Producer<CommandId, Command> commandProducer;
+
+  public CommandTopic(
+      final String commandTopic,
+      final Map<String, Object> commandConsumerProperties
+  ) {
+    this(
+        new KafkaConsumer<>(
+        commandConsumerProperties,
+        CommandTopic.getJsonDeserializer(CommandId.class, true),
+        CommandTopic.getJsonDeserializer(Command.class, false)
+    ),
+
+    new KafkaProducer<>(
+        commandConsumerProperties,
+        CommandTopic.getJsonSerializer(true),
+        CommandTopic.getJsonSerializer(false)
+    ));
+    commandConsumer.assign(Collections.singleton(new TopicPartition(commandTopic, 0)));
+  }
+
+  public CommandTopic(
+      final Consumer<CommandId, Command> commandConsumer,
+      final Producer<CommandId, Command> commandProducer
+  ) {
+    this.commandConsumer = commandConsumer;
+    this.commandProducer = commandProducer;
+  }
+
+
+  @SuppressWarnings("unchecked")
+  public void send(final ProducerRecord producerRecord)
+      throws ExecutionException, InterruptedException {
```

Comment:
These exceptions cause the internal implementation details to bleed out. We discussed this if you remember. Better to unwrap the `ExecutionException`.

Given that you should also be encapsulating the command topic name inside of this class, this method should become something like:

```
public void send(final CommandId commandId, final Command command) throws Exception {
    try {
      commandProducer.send(new ProducerRecord<>(commandTopicName, commandId, command)).get();
    } catch (final ExecutionException e) {
      if (e.getCause() instanceof Exception) {
        throw (Exception)e.getCause();
      }
      throw new RuntimeException(e.getCause());
    }
  }
```